pkg/project/auth/watch.go

```diff
@@ -0,0 +1,217 @@
+package auth
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/golang/glog"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/auth/user"
+	"k8s.io/kubernetes/pkg/storage/etcd"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/watch"
+
+	projectapi "github.com/openshift/origin/pkg/project/api"
+	projectcache "github.com/openshift/origin/pkg/project/cache"
+)
+
+type CacheWatcher interface {
+	// GroupMembershipChanged is called serialially for all changes for all watchers.  This method MUST NOT BLOCK.
+	// The serial nature makes reasoning about the code easy, but if you block in this method you will doom all watchers.
+	GroupMembershipChanged(namespaceName string, latestUsers, latestGroups, removedUsers, removedGroups, addedUsers, addedGroups sets.String)
+}
+
+type WatchableCache interface {
+	// RemoveWatcher removes a watcher
+	RemoveWatcher(CacheWatcher)
+	// List returns the set of namespace names the user has access to view
+	List(userInfo user.Info) (*kapi.NamespaceList, error)
+}
+
+// userProjectWatcher converts a native etcd watch to a watch.Interface.
+type userProjectWatcher struct {
+	username string
+	groups   []string
+
+	// cacheIncoming is a buffered channel used for notification to watcher.  If the buffer fills up,
+	// then the watcher will be removed and the connection will be broken.
+	cacheIncoming chan watch.Event
+	// cacheError is a cached channel that is put to serialially.  In theory, only one item will
+	// ever be placed on it.
+	cacheError chan error
+
+	// outgoing is the unbuffered `ResultChan` use for the watch.  Backups of this channel will block
+	// the default `emit` call.  That's why cacheError is a buffered channel.
+	outgoing chan watch.Event
+	// userStop lets a user stop his watch.
+	userStop chan struct{}
+
+	// stopped prevents closing closed channels
```

Comment:
Why are booleans preffered for this type of job instead of just checking if the channel is closed before closing it ie. try to receive from it?